### PR TITLE
implement floor division

### LINKED ISSUE #139: implement floordiv
In most cases a floor division can be obtained via arb. This would avoid the division in the number field.

### PR DESCRIPTION
**Pull request by saraedum**
fixes #139

### CODE REVIEW

**Comment by saraedum** [comment]
@videlec any objections to merging this? I'd like to cut a new rc release after that.

**Comment by videlec** [comment]
We could completely avoid the `renf_elem` division `b / c` here. Namely
- refine `quotient = arb(b) / arb(c)` until its diameter is < 0.5
- if `quotient` does not contain an integer: easy
- if `quotient` contains the integer `n` then if `n*c <= b` return `n` otherwise return `n-1`
Above you only perform an integer multiplication which is way cheaper in arithmetic than the plain division.

**Comment by saraedum** [comment]
@videlec that's much smarter of course. It's implemented like that now. The tests took 6s before, now they take 4s.

File: renf_elem/fdiv.c [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        slong bcond = WORD_MIN;
+        slong ccond = WORD_MIN;
+        slong prec;
+
+        prec = FLINT_MAX(nf->prec, arb_rel_accuracy_bits(nf->emb));
+        renf_elem_relative_condition_number_2exp(&bcond, b, nf);
+        renf_elem_relative_condition_number_2exp(&ccond, c, nf);
+
+        renf_elem_set_evaluation(b, nf, prec + bcond);
+        renf_elem_set_evaluation(c, nf, prec + bcond);
```
Comment (videlec) [inline renf_elem/fdiv.c:29]: Shouldn't it be ccond here?

File: renf_elem/fdiv.c [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+        /* ensure that the encolsure is not infinite */
+        while (!arb_is_finite(b->emb))
+        {
+            prec *= 2;
+            renf_refine_embedding(nf, prec);
+            renf_elem_set_evaluation(b, nf, prec + bcond);
+        }
+
```
Comment (videlec) [inline renf_elem/fdiv.c:38]: Shouldn't we ensure that c->emb does not contain zero?

**Review by videlec — CHANGES_REQUESTED** [review]

File: renf_elem/fdiv.c (hunk 1 above)
Comment (saraedum) [reply renf_elem/fdiv.c]: oops

File: renf_elem/fdiv.c (hunk 2 above)
Comment (saraedum) [reply renf_elem/fdiv.c:37]: more oops

File: renf_elem/fdiv.c [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        renf_elem_set_evaluation(c, nf, prec + ccond);
+
+        /* ensure that the enclosure of the divisor does not contain zero */
+        while (!arb_is_nonzero(c->emb))
+        {
+            prec *= 2;
+            renf_refine_embedding(nf, prec);
+            renf_elem_set_evaluation(c, nf, prec + ccond);
+        }
+
```
Comment (saraedum) [inline renf_elem/fdiv.c:47]: If the precision has changed for c, I could now call `renf_elem_set_evaluation` for `b` again. But it's not clear that it's always a good idea.

File: renf_elem/fdiv.c (hunk 2 above)
Comment (videlec) [reply renf_elem/fdiv.c:37]: I don't see how tests could catch such thing. The `ccond` is just a number to ensure that you end up with an embedding with at least `prec` valid bits. Did you have something in mind?

File: renf_elem/fdiv.c (hunk 3 above)
Comment (videlec) [reply renf_elem/fdiv.c:47]: I think it is fine the way you wrote it. In many situations the default embeddings are already good enough to determine the fdiv.

File: renf_elem/fdiv.c (hunk 2 above)
Comment (saraedum) [reply renf_elem/fdiv.c:37]: No, I mean, why the tests did not catch the `c` contains zero case. Apparently, the random renf_elems do not contain numbers very close to zero.

Comment (videlec) [reply renf_elem/fdiv.c:37]: Oh, true. This is annoying. However creating a small nonzero number field element is not trivial. If you pick coefficients at random you will never find a small one. Maybe we should tweak the random generator to make them often enough. see #143

**Review by videlec — APPROVED** [review]

**Pull request merged**